File: fun/aoc2022/day13.py
import json
from functools import cmp_to_key
from math import prod

example_data = """
[1,1,3,1,1]
[1,1,5,1,1]

[[1],[2,3,4]]
[[1],4]

[9]
[[8,7,6]]

[[4,4],4,4]
[[4,4],4,4,4]

[7,7,7,7]
[7,7,7]

[]
[3]

[[[]]]
[[]]

[1,[2,[3,[4,[5,6,7]]]],8,9]
[1,[2,[3,[4,[5,6,0]]]],8,9]

""".strip()


INPUT_FILE_PATH = "fun/aoc2022/input_files/day13_gap.txt"


def read_file():
    with open(INPUT_FILE_PATH, "r") as f:
        return f.read().strip()


def solution(data: str, part=1):
    if part == 1:
        pairs = data.split("\n\n")
        r = {}
        for i, pair in enumerate(pairs):
            L1, L2 = list(map(json.loads, pair.split("\n")))
            n = check_order(L1, L2)
            r[i + 1] = n
        return sum([k for k, v in r.items() if v == 1])

    if part == 2:
        lines = data.replace("\n\n", "\n").split("\n")
        lines = list(map(json.loads, lines))
        dividers = [[[2]], [[6]]]
        lines += dividers
        s = sorted(lines, key=cmp_to_key(check_order), reverse=True)
        return prod(s.index(i) + 1 for i in dividers)


def check_order(L1, L2):
    """
    Return: 
        1 for correct order
        0 for tie
        -1 for incorrect order
    """
    l1 = iter(L1)
    l2 = iter(L2)

    lstop = False
    rstop = False
    try:
        left = next(l1)
    except StopIteration:
        lstop = True
    try:
        right = next(l2)
    except StopIteration:
        rstop = True

    if lstop and not rstop:
        return 1
    elif lstop and rstop:
        return 0
    elif rstop and not lstop:
        return -1

    if isinstance(left, int) and isinstance(right, int):
        if left < right:
            return 1
        elif right < left:
            return -1
        elif right == left:
            return check_order(l1, l2)

    elif isinstance(left, int):
        left = [left]
    elif isinstance(right, int):
        right = [right]

    val = check_order(left, right)
    if val != 0:
        return val
    else:
        return check_order(l1, l2)


if __name__ == "__main__":
    assert solution(example_data, part=1) == 13
    print("Part 1: ", solution(read_file(), part=1))

    assert solution(example_data, part=2) == 140
    print("Part 2: ", solution(read_file(), part=2))

File: fun/aoc2022/day13.jl
function solution(ifile; part)
    s = 0
    data = read(ifile, String)
    if part == 1
        pairs = split(data, "\n\n")
        for i ∈ eachindex(pairs)
            lines = split(pairs[i], "\n")

            # read signals
            x = eval(Meta.parse(lines[1]))
            y = eval(Meta.parse(lines[2]))

            # determine if in right order
            check = in_order(x, y)
            if check ≥ 0
                s += i
            end
        end
        return s
    elseif part == 2
        lines = eval.(Meta.parse.(split(replace(data, "\n\n" => "\n"), "\n")))
        dividers = [[[2]], [[6]]]
        for d ∈ dividers
            push!(lines, d)
        end
        s = sort(lines, lt=in_order_TF)
        return prod(findfirst(isequal(d), s) for d ∈ dividers)
    end
end

"""
    in_order(x, y)

 1 - correct order
 0 - tie
-1 - incorrect order
"""
function in_order(x, y)
    for i ∈ eachindex(x)
        # x is longer than y → incorrect
        if i > length(y)
            return -1
        end

        # compare ith elements of x and y
        left = x[i]
        right = y[i]

        # deal with possible types
        if typeof(left) == typeof(right) <: Integer
            # regular integer ordering
            if left < right
                return 1
            elseif left > right
                return -1
            else
                # equal, go to next element
                continue
            end
        elseif typeof(left) <: Integer
            # right must be a list, change left
            left = [left]
        elseif typeof(right) <: Integer
            # left must be a list, change right
            right = [right]
        end

        # recursively run check on left and right
        check = in_order(left, right) 
        if check != 0
            return check
        end
    end

    # y is longer than x → correct
    if length(y) > length(x)
        return 1
    end

    # otherwise, tie
    return 0
end

function in_order_TF(x, y)
    return in_order(x, y) == 1
end

println(solution("input_files/day13_hgp.txt"; part=1)) # 0.2 s
println(solution("input_files/day13_hgp.txt"; part=2)) # 0.2 s